test/e2e/kvc_test.go

```diff
@@ -0,0 +1,370 @@
+package e2e
+
+import (
+	"flag"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	"github.com/stretchr/testify/require"
+
+	crv1 "github.com/kubeflow/experimental-kvc/pkg/apis/kvc/v1"
+	crv1_client "github.com/kubeflow/experimental-kvc/pkg/client/clientset/versioned"
+	crv1_volume_manager "github.com/kubeflow/experimental-kvc/pkg/client/clientset/versioned/typed/kvc/v1"
+	"github.com/kubeflow/experimental-kvc/pkg/states"
+	"github.com/kubeflow/experimental-kvc/pkg/util"
+)
+
+var namespace = flag.String("namespace", "e2e-test", "namespace used for the e2e test")
+
+func makeClients(t *testing.T) (crv1_volume_manager.VolumeManagerInterface, *kubernetes.Clientset) {
+	user, err := user.Current()
+	require.Nil(t, err)
+
+	config, err := util.BuildConfig(filepath.Join(user.HomeDir, ".kube/config"))
+	require.Nil(t, err)
+
+	k8sClient, err := kubernetes.NewForConfig(config)
+	require.Nil(t, err)
+	require.NotNil(t, k8sClient)
+
+	crdClient, err := crv1_client.NewForConfig(config)
+	require.Nil(t, err)
+	require.NotNil(t, crdClient)
+
+	return crdClient.KvcV1().VolumeManagers(*namespace), k8sClient
+}
+
+func makeVolumeManager(volumeConfigs []crv1.VolumeConfig) *crv1.VolumeManager {
+	name := fmt.Sprintf("kvc-e2e-test-%s", uuid.NewUUID())
+	return &crv1.VolumeManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: crv1.VolumeManagerSpec{
+			VolumeConfigs: volumeConfigs,
+			State:         states.Running,
+		},
+	}
+}
+
+// WaitForCRState polls for an expected CR state untill it reaches a timeout.
+func waitForCRState(crdClient crv1_volume_manager.VolumeManagerInterface, name string, state states.State) error {
+	return waitPoll(func() (bool, error) {
+		volman, err := crdClient.Get(name, metav1.GetOptions{})
+		if err == nil && volman.Status.State == state {
+			return true, nil
+		}
+
+		return false, err
+	})
+}
+
+func waitPoll(waitFunc func() (bool, error)) error {
+	return wait.Poll(1*time.Second, 30*time.Second, waitFunc)
+}
+
+func TestVolumeManager(t *testing.T) {
+	crdClient, _ := makeClients(t)
+
+	testCases := []struct {
+		description   string
+		volumeConfigs []crv1.VolumeConfig
+		expSuccess    bool
+		expError      string
+		expHP         bool
+		expNA         bool
+		expPVC        bool
+	}{
+		// Positive test cases.
+		{
+			description: "single vc - S3 - no error",
+			volumeConfigs: []crv1.VolumeConfig{
+				{
+					ID:          "vol1",
+					Replicas:    1,
+					SourceType:  "S3",
+					SourceURL:   "s3://e2e-test/cifar-100-python.tar.gz",
+					EndpointURL: "http://10.63.255.0:9000",
```

Comment:
If the service restarts/changes, we might have to change the IP here. I think we should convert this to an argument too.

Reply:
I converted this to an arg.